File: pos-maui/Modelo/Dao/UsuarioDao.cs
using Dapper;
using PosMaui.Modelo.Entidad;
using System.Security.Cryptography;
using System.Text;

namespace PosMaui.Modelo.Dao;

public static class UsuarioDao
{
    public static List<Usuario> ObtenerActivos()
    {
        using var conn = ConexionBD.Abrir();
        return conn.Query<Usuario>("""
            SELECT u.id AS Id, u.nombre AS Nombre, r.nombre AS Rol, u.activo AS Activo
            FROM usuarios u
            JOIN roles r ON r.id = u.rol_id
            WHERE u.activo = true
            ORDER BY u.nombre
            """).ToList();
    }

    public static Usuario? VerificarPin(string nombre, string pin)
    {
        string hash = Sha256(pin);
        using var conn = ConexionBD.Abrir();
        return conn.QueryFirstOrDefault<Usuario>("""
            SELECT u.id AS Id, u.nombre AS Nombre, r.nombre AS Rol, u.activo AS Activo
            FROM usuarios u
            JOIN roles r ON r.id = u.rol_id
            WHERE u.nombre = @Nombre AND u.pin_hash = @Hash AND u.activo = true
            """, new { Nombre = nombre, Hash = hash });
    }

    public static void Crear(string nombre, string pin, int rolId)
    {
        using var conn = ConexionBD.Abrir();
        conn.Execute("""
            INSERT INTO usuarios (nombre, pin_hash, rol_id)
            VALUES (@Nombre, @Hash, @RolId)
            """, new { Nombre = nombre, Hash = Sha256(pin), RolId = rolId });
    }

    public static void ToggleActivo(int id, bool activo)
    {
        using var conn = ConexionBD.Abrir();
        conn.Execute("UPDATE usuarios SET activo = @Activo WHERE id = @Id",
                     new { Activo = activo, Id = id });
    }

    public static string Sha256(string input)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(input));
        return Convert.ToHexString(bytes).ToLower();
    }
}

File: pos-python/db/usuario_dao.py
import hashlib
from db.connection import DB


def _sha256(text: str) -> str:
    return hashlib.sha256(text.encode()).hexdigest()


def verificar_pin(nombre: str, pin: str):
    """Returns user dict or None."""
    return DB.fetchone(
        """SELECT u.id, u.nombre, r.nombre AS rol
           FROM usuarios u
           JOIN roles r ON r.id = u.rol_id
           WHERE u.nombre = %s AND u.pin_hash = %s AND u.activo = TRUE""",
        (nombre, _sha256(pin))
    )


def find_all():
    return DB.fetchall(
        """SELECT u.id, u.nombre, r.nombre AS rol, u.activo, u.creado_en
           FROM usuarios u JOIN roles r ON r.id = u.rol_id
           ORDER BY u.nombre"""
    )


def find_roles():
    return DB.fetchall("SELECT id, nombre FROM roles ORDER BY nombre")


def insert(nombre: str, pin: str, rol_id: int):
    DB.execute(
        "INSERT INTO usuarios (nombre, pin_hash, rol_id) VALUES (%s, %s, %s)",
        (nombre, _sha256(pin), rol_id)
    )


def cambiar_pin(usuario_id: int, nuevo_pin: str):
    DB.execute(
        "UPDATE usuarios SET pin_hash = %s WHERE id = %s",
        (_sha256(nuevo_pin), usuario_id)
    )


def toggle_activo(usuario_id: int):
    DB.execute(
        "UPDATE usuarios SET activo = NOT activo WHERE id = %s",
        (usuario_id,)
    )
